### PR TITLE
Add QPY support for `switch` statement

### PR DESCRIPTION
**Pull request by jakelishman**
### Summary

QPY had a somewhat implicit assumption that most classes can be reconstructed by doing `type(obj)(*obj.params)`, which `SwitchCaseOp` very much does not abide by - to do so would compromise more core parts of how it is meant to function.  This commit puts in a special case so that instances of `SwitchCaseOp` present their "params" as the information necessary to reconstruct the op in the above form, rather than actually using the parameters.  That didn't feel _entirely_ satisfying, but does work fine.

### CODE REVIEW

File: qiskit/qpy/type_keys.py
Comment (jakelishman) [reply qiskit/qpy/type_keys.py:104]: I can move it if you think there's something more sensible for me to put it - I wasn't sure.  It's unlikely that control flow itself will gain more keys that need adding here, but we will need an entire new subtree to handle extended classical expressions (Qiskit/rfcs#31) at some point in the future (that's completely separate to this).  It's likely they'll be able to have an entirely separate namespace, though.

I'm not sure what advantage there is to having these separated enums; there's cross-dependencies between them already, so it's insufficient to just make a key unique within a single enum.  In the general case, the keys have to be globally unique.  I found that when trying to use an `r` key in `Value`: it clashed with an `r` for the Python builtin `range` in `Container`.  I feel like it might be cleaner to unify these into a global enumeration and enforce complete uniqueness; once we've got the enum, there's no real need for the ASCII characters to have any mnemonic value, and since the allowed keys in any given place in the QPY format are often a mixture of a few enums, they're not offering much type safety at the moment.

Comment (jakelishman) [reply qiskit/qpy/type_keys.py:104]: That's fair enough, yeah.  I'm not really familiar with the QPY code at all, so I'm happy to put it somewhere else if you'd prefer.

We currently don't have any plans to extend the control-flow handling, though I suppose at some very indeterminate time in the future there _might_ be the concept of a relocatable subroutine call.  That's probably a year+ away at least, though, so I wouldn't worry about it.

File: qiskit/qpy/common.py
Comment (nkanazawa1989) [inline qiskit/qpy/common.py:23]: You don't need to bump QPY version since you didn't break any data structure.

File: qiskit/qpy/__init__.py
Comment (nkanazawa1989) [inline qiskit/qpy/__init__.py:134]: This block can go to 
https://github.com/Qiskit/qiskit-terra/blob/912ba0783420463b4adf7f53bee7cba04bd915f8/qiskit/qpy/__init__.py#L905-L919

maybe we need better document structure.

**Review by nkanazawa1989 — CHANGES_REQUESTED** [review]
Thanks Jake. I reviewed the PR and the proposed change looks good to me.

File: qiskit/qpy/__init__.py
Comment (jakelishman) [reply qiskit/qpy/__init__.py]: My problem was that that block is under the "version 1" heading, and the new type keys aren't defined in version 1, so it didn't seem right to put it there.  For example, in version 4, we add two type keys to the same `INSTRUCTION_PARAMS` struct, but don't put them in version 1.

File: qiskit/qpy/common.py
Comment (jakelishman) [reply qiskit/qpy/common.py]: It feels odd not to, to me - since I defined some type keys that weren't in previous versions, we'd be breaking that property that all things written to the original version 6 spec would be able to load these new programs.  That said, the point's moot since your PR absolutely needed a version bump.

File: qiskit/qpy/__init__.py
Comment (mtreinish) [reply qiskit/qpy/__init__.py]: Yeah, I think we need to rework the documentation. At the time it was trying to build off what each version's structs looked like so you'd incrementally build the formats off each other. I think we probably need to reorganize it so the current format is at the top and then we describe (with full examples what things looked like in older versions). I'll open an issue about this so we can track it as something in a follow up. For this we can probably leave the new addition in just the version 7 section and fix this in that followup.

**Review by mtreinish — APPROVED** [review]
Overall this LGTM, it's a pretty straightforward addition to handle the extra edge cases around the switch statement. But I'd like to get a final ack from @nkanazawa1989 before enqueuing for merge to ensure his concerns are addressed.

File: qiskit/qpy/common.py
Comment (jlapeyre) [reply qiskit/qpy/common.py]: If there is any possibility of a bug being introduced, even if it's not obvious, a version bump would help consumers track it down.

**Review by nkanazawa1989 — APPROVED** [review]
LGTM. My comment above just followed precedent and I also tend to prefer bumping version with new data keys. Reorganizing documentation and adding some policy for version control would be really nice.

**Pull request merged**
